entrypoint.sh

```diff
@@ -4,21 +4,61 @@ set -e
 PG_CONFIG=/etc/pgbouncer/pgbouncer.ini
 PG_USER=postgres
 
-# Generate key and certificate for ssl connection
-if [ ! -r /etc/pgbouncer/ssl/server.key ]; then
-  echo "Generating new key and certificate for SSL connection"
-  CWD=`pwd`
-  cd /etc/pgbouncer/ssl/
-  openssl genrsa -des3 -passout pass:xxxx -out server.pass.key 2048
-  openssl rsa -passin pass:xxxx -in server.pass.key -out server.key
-  rm -f server.pass.key
-  openssl req -new -key server.key -out server.csr -subj "${CERTIFICATE_DESCRIPTION}"
-  openssl x509 -req -in server.csr -signkey server.key -out server.crt
-  rm -f server.csr
-  chown -R ${PG_USER}:${PG_USER} /etc/pgbouncer/ssl/
-  cd "$CWD"
+mkdir -p /etc/pgbouncer/ssl
+
+if [ -n "${CA_KEY}" ]
+then
+    echo "${CA_KEY}" > /etc/pgbouncer/ssl/root.key
+else
+    if [ -r "${CA_KEY_FILENAME}" ]
+    then
+       cp "${CA_KEY_FILENAME}" /etc/pgbouncer/ssl/root.key
+    else
+        echo "You must provide the CA Certificate and CA root key as environment variable. Check README.md"
+        exit 1
+    fi
+fi
+
+if [ -n "${CA_CERT}" ]
+then
+    echo "${CA_CERT}" > /etc/pgbouncer/ssl/root.crt
+else
+    if [ -r "${CA_CERT_FILENAME}" ]
+    then
+        cp "${CA_CERT_FILENAME}" /etc/pgbouncer/ssl/root.crt
+    else
+        echo "You must provide the CA Certificate and CA root key as environment variable. Check README.md"
+        exit 1
+    fi
+fi
+
+if [ -n "${SERVER_KEY_SUBJECT}" ]
+then
+   export SERVER_KEY_SUBJECT="/C=CH/ST=Vaud/L=Lausanne/O=Camptocamp S.A./OU=BS/CN=pg/emailAddress=contact@camptocamp.com/"
 fi
 
+if [ ! \( -r /opt/pgbouncer/ssl/server.key -a  -r /opt/pgbouncer/ssl/server.crt \) ]
+then
+    echo "Generating new server key for " ${SERVER_KEY_SUBJECT}
+    openssl req -new -nodes -text -out /opt/pgbouncer/ssl/server.csr -keyout /opt/pgbouncer/ssl/server.key -subj "${SERVER_KEY_SUBJECT}"
+    echo "Signing key with root certificate"
+    ls -l /opt/pgbouncer/ssl/server.csr
+    ls -l /opt/pgbouncer/ssl/server.key
+    ls -l /etc/pgbouncer/ssl/root.crt
+    ls -l /etc/pgbouncer/ssl/root.key
+    cat /etc/pgbouncer/ssl/root.crt
+    openssl x509 -req -in /opt/pgbouncer/ssl/server.csr -text -days 365 -CA /etc/pgbouncer/ssl/root.crt -CAkey /etc/pgbouncer/ssl/root.key -CAcreateserial -out /opt/pgbouncer/ssl/server.crt
+fi
+cp /opt/pgbouncer/ssl/server.key /opt/pgbouncer/ssl/server.crt /etc/pgbouncer/ssl/
+chmod 600 /etc/pgbouncer/ssl/root.crt /etc/pgbouncer/ssl/root.key
+chmod 600 /etc/pgbouncer/ssl/server.key /etc/pgbouncer/ssl/server.crt
+
+chown ${PG_USER} /etc/pgbouncer/ssl/server.key /etc/pgbouncer/ssl/server.crt /etc/pgbouncer/ssl/root.crt /etc/pgbouncer/ssl/root.key
```

Comment:
How can I check if this is needed? Because the comment you point out mentions this being done for postgresql's `initdb` command, and there are all kind of things related to ident authentication, and checking the the user's personal database in postgresql which I don't think apply for pgbouncer.

Reply:
If you try to use `psql` command you will need it. I agree with you, maybe we don't need it. You can leave it as is for now. I will check later on a openshift platform.